### PR TITLE
opt(server): If unspecified, use num_shards == proactor_threads

### PR DESCRIPTION
**Pull request by chakaz**

### CODE REVIEW

File: src/server/main_service.cc [hunk 1]
```diff
@@ -661,9 +661,7 @@ void Service::Init(util::AcceptServer* acceptor, std::vector<facade::Listener*>
   pp_.Await([reg](uint32_t index, ProactorBase* pb) { ServerState::Init(index, reg); });
 
   uint32_t shard_num = GetFlag(FLAGS_num_shards);
-  if (shard_num == 0) {
-    shard_num = pp_.size() > 1 ? pp_.size() - 1 : pp_.size();
-  } else if (shard_num > pp_.size()) {
```
Comment (chakaz) [inline src/server/main_service.cc:666]: BTW, we might want to `exit(-1)`, or at least print a `WARNING` in this case?

Comment (romange) [reply src/server/main_service.cc:666]: warning is fine

**Comment by romange** [comment]
@chakaz  I opened https://github.com/romange/helio/pull/153 to address the failure.

**Comment by romange** [comment]
I think it's not enough - we should also fix regression tests

File: tests/dragonfly/instance.py [hunk 2]
```diff
@@ -74,6 +74,12 @@ def __init__(self, params: DflyParams, args):
             del self.args["logtostderr"]
             self.args["alsologtostderr"] = None
 
+        # Run with num_shards = (proactor_threads - 1) if possible, so help expose bugs
+        if "proactor_threads" in self.args and "num_shards" not in self.args:
```
Comment (chakaz) [inline tests/dragonfly/instance.py:78]: Now we only set `num_shards` when `proactor_threads` is set, but it's probably unset in most cases. The issue is figuring out the unset value.
Shall I apply this logic even if `proactor_threads` is not set? If so, to which value?

**Comment by chakaz** [comment]
@romange kindly take another look :pray:

File: tests/dragonfly/instance.py (hunk 2 above)
Comment (romange) [reply tests/dragonfly/instance.py:78]: if not set, please set it to `psutil.cpu_count()` and num_shards to -1 if the former is `>2` @chakaz

Comment (chakaz) [reply tests/dragonfly/instance.py]: done

**Review by romange — APPROVED** [review]

File: src/server/main_service.cc [hunk 3]
```diff
@@ -686,9 +686,10 @@ void Service::Init(util::AcceptServer* acceptor, std::vector<facade::Listener*>
   pp_.Await([reg](uint32_t index, ProactorBase* pb) { ServerState::Init(index, reg); });
 
   uint32_t shard_num = GetFlag(FLAGS_num_shards);
-  if (shard_num == 0) {
-    shard_num = pp_.size() > 1 ? pp_.size() - 1 : pp_.size();
```
Comment (adiholden) [inline src/server/main_service.cc:690]: @chakaz can you share with me what was the reason for this fix?

Comment (chakaz) [reply src/server/main_service.cc:690]: Sure, when we benchmarked bullmq I did runs with num_shards == threads in addition to num_shards == (threads - 1), and it was consistently better like this.

**Pull request merged**